### ISSUE TITLE
Display Status bars on Touch bar (Macbook Pro)

### ISSUE THREAD

**Issue by nlruk** [runelite/runelite#11584]
Hello,

Since you are actively supporting the OSX ecosystem could you consider making a plugin that displays the Status bars and other information like notifications and timers on the Touch Bar (Macbook Pro)

https://github.com/Thizzer/JTouchBar

![emoji](https://user-images.githubusercontent.com/63312904/81915530-9e4cff00-95ca-11ea-9815-5eefcc3af3b7.png)
<img width="787" alt="Screenshot 2020-05-14 at 10 31 05" src="https://user-images.githubusercontent.com/63312904/81918203-1832b780-95ce-11ea-9fef-4eb768ba39db.png">

![emoji game](https://user-images.githubusercontent.com/63312904/81915538-a3aa4980-95ca-11ea-8d2e-207244e075aa.jpg)
![eemoji pac](https://user-images.githubusercontent.com/63312904/81915553-a86efd80-95ca-11ea-849e-e9f8056b3422.jpg)

**Comment by while-loop** [comment]
This is a pretty cool idea, but I'm not too sure how useful it'll be. 

1. Idk if remapping shortcuts to the touchbar would be allowed (ie opening the invy, or activating quick prayers)
2. Duplicating things that are already onscreen seems like a waste of space on the touchbar since runelite has to be focused to show the touchbar anyway (hp, prayer, timers)
3. Low-key think that repo/lib is dead.. at a quick glance. 

But for things client related, I could see a screenshot button useful. Or opening the side panels.

**Comment by Broooklyn** [comment]
While it's not integrated into runelite the way you'd like, take a look at https://github.com/pg8wood/OSRS-touch-bar 

I like the idea of full (not OP) integration though.

**Comment by nlruk** [comment]
> While it's not integrated into runelite the way you'd like, take a look at https://github.com/pg8wood/OSRS-touch-bar
> 
> I like the idea of full (not OP) integration though.

Would that be difficult to make it an external plugin?

**Comment by makingcomment** [comment]
I think @TylerADavis has a branch where some touchbar support is implemented and/or a WIP, but he hasn't updated it in a couple months.

**Comment by TylerADavis** [comment]
Yeah, I started a branch a while back and I successfully got buttons displayed on the Touch Bar and reacting to touch events. The original intent was to just have buttons with icons to replace the f1 f2 etc. shortcuts for opening inventory and prayer and such. I think this would be fine since you can already do it with function keys, and functionally this isn’t any different.

It’s definitely doable, but I’m not sure what the best things to put on the Touch Bar would be. I don’t think timers would be great necessarily, since you would need to look at your hands instead of the display, but it sounds more exciting than buttons for ui sections.

I’d run into three issues

1) I wasn’t sure where to find the actual assets for the different tab icons, so I was just using the ones from the ui styles plugin
2) the touchbar buttons were too large to fit all ui tab buttons on the touch at at once
3) I wasn’t sure how to switch between ui tabs after receiving the button push

I did it with jtouchbar and it seemed to work alright.

**Comment by nlruk** [comment]
> It’s definitely doable, but I’m not sure what the best things to put on the Touch Bar would be. I don’t think timers would be great necessarily since you would need to look at your hands instead of the display, but it sounds more exciting than buttons for UI sections.

I'm sure amongst all of us we can find some good ideas.

- Skill XP tracker

- World switcher

- Minimize Chat

- Home Teleport

**Comment by nlruk** [comment]
Would GrandExchangeOffersPanel progress bars be possible to be displayed on the touch bar?

**Comment by pg8wood** [comment]
Howdy, I'm the author of OSRS Touch Bar. I've been thinking about ways to add Touch Bar functionality to RuneLite for awhile now. JTouchBar seemed promising when it came out, but I'd be hesitant to use it in 2020, as it looks like it's pretty much been abandoned. 

I plan on looking into the feasibility of adding a small Cocoa wrapper to RuneLite on macOS. If my thought process is correct, RuneLite would look the same to the users, but we'd be able to leverage Touch Bar and other Cocoa APIs directly rather than relying on brittle Java/Objective-C bridges.

I think being able to develop the Touch Bar natively using Cocoa is key. I don't see any major Touch Bar work being feasible using JTouchBar in its current state. Even porting my F-key mappings from OSRS Touch Bar felt _very_ hacky in JTouchBar.

If I'm able to add the macOS application layer to RuneLite easily, I'd be happy to discuss next steps re: Touch Bar with the RuneLite devs. Before I start that conversation though, I'd want to make sure this Cocoa wrapper would be a light touch. I wouldn't imagine anyone would want to majorly change the way RuneLite runs on macOS for a few niche features.

**Comment by Adam-** [comment]
> 
> 
> Howdy, I'm the author of OSRS Touch Bar. I've been thinking about ways to add Touch Bar functionality to RuneLite for a while now. JTouchBar seemed promising when it came out, but I'd be hesitant to use it in 2020, as it looks like it's pretty much been abandoned.
> 
> I plan on looking into the feasibility of adding a small Cocoa wrapper to RuneLite on macOS. If my thought process is correct, RuneLite would look the same to the users, but we'd be able to leverage Touch Bar and other Cocoa APIs directly rather than relying on brittle Java/Objective-C bridges.
> 
> I think being able to develop the Touch Bar natively using Cocoa is key. I don't see any major Touch Bar work being feasible using JTouchBar in its current state. Even porting my F-key mappings from OSRS Touch Bar felt _very_ hacky in JTouchBar.
> 
> If I'm able to add the macOS application layer to RuneLite easily, I'd be happy to discuss next steps re: Touch Bar with the RuneLite devs. Before I start that conversation though, I'd want to make sure this Cocoa wrapper would be a light touch. I wouldn't imagine anyone would want to majorly change the way RuneLite runs on macOS for a few niche features.

A good way to do this may be to make a plugin which exposes an api locally that the touchbar app could talk to? I have https://github.com/Adam-/runelite-plugins/tree/http-plugin which some people use for displaying their hitpoints/prayer in their taskbar on Linux which is similar. Only this would have to be a 2-way thing I guess.

**Comment by pg8wood** [comment]
Neat, thanks for the resource. I'm less familiar with the RuneLite plugin ecosystem, but that definitely looks like a good avenue to explore. It gets a bit tricky due to how the OS handles the Touch Bar presentation alongside the focused application window, but I'll check it out!

**Comment by pg8wood** [comment]
![image](https://user-images.githubusercontent.com/16855664/83456146-26d1f900-a42d-11ea-8fff-b0cfcfc4a0d8.png)
Update: I've got a rough prototype of a plugin-based approach working. I'm investigating bundling a windowless native macOS app that the plugin will launch and talk to (pretty similar to the taskbar plugins on Linux). 

Working on getting things a little more solid foundation-wise 🙂

**Comment by TylerADavis** [comment]
One minor comment, I believe the esc key is a shortcut for the inventory, but it's currently replaced by that X symbol. Is there a way to include a shortcut for inventory as well?

Perhaps it could be omitted on laptops with second gen touchbars, as they have permanent escape keys.

**Comment by pg8wood** [comment]
> One minor comment, I believe the esc key is a shortcut for the inventory, but it's currently replaced by that X symbol. Is there a way to include a shortcut for inventory as well?
> 
> 
> 
> Perhaps it could be omitted on laptops with second gen touchbars, as they have permanent escape keys.

Looks like it’s easily doable! Added it to the branch I’m working on 👍

**Comment by Optumasfx** [comment]
@pg8wood any updates on development? this is so exciting for mac users 👍

**Comment by pg8wood** [comment]
👋 Hey @Optumasfx admittedly no--2020 started getting pretty crazy around the time I last looked at this. Thanks for the reminder! I'll bump it up a few slots on my TODO list 😃

**Comment by stoddy15** [comment]
> 👋 Hey @Optumasfx admittedly no--2020 started getting pretty crazy around the time I last looked at this. Thanks for the reminder! I'll bump it up a few slots on my TODO list 😃

wow awesome brotherrrrrr
